src/fam.rs

```diff
@@ -0,0 +1,623 @@
+    ///
+    /// The caller must ensure that the fam_struct outlives the pointer this
+    /// function returns, or else it will end up pointing to garbage.
+    ///
+    /// Modifying the container referenced by this pointer may cause its buffer
+    /// to be reallocated, which would also make any pointers to it invalid.
+    ///
+    pub fn as_fam_struct_ptr(&self) -> *const T {
+        self.as_fam_struct_ref()
+    }
+
+    /// Get a mutable pointer to the [`FamStruct`](trait.FamStruct.html) instance.
+    ///
+    /// The caller must ensure that the fam_struct outlives the pointer this
+    /// function returns, or else it will end up pointing to garbage.
+    ///
+    /// Modifying the container referenced by this pointer may cause its buffer
+    /// to be reallocated, which would also make any pointers to it invalid.
+    ///
+    pub fn as_mut_fam_struct_ptr(&mut self) -> *mut T {
+        self.as_mut_fam_struct()
+    }
+
+    /// Get the elements slice.
+    ///
+    pub fn as_slice(&self) -> &[T::Entry] {
+        self.as_fam_struct_ref().as_slice()
+    }
+
+    /// Get the mutable elements slice.
+    ///
+    pub fn as_mut_slice(&mut self) -> &mut [T::Entry] {
+        self.as_mut_fam_struct().as_mut_slice()
+    }
+
+    /// Get the number of elements of type `FamStruct::Entry` currently in the vec.
+    ///
+    fn len(&self) -> usize {
+        self.as_fam_struct_ref().len()
+    }
+
+    /// Get the capacity of the `FamStructWrapper` measured in elements of type `FamStruct::Entry`.
+    ///
+    fn capacity(&self) -> usize {
+        FamStructWrapper::<T>::fam_len(self.mem_allocator.capacity())
+    }
+
+    /// Reserves capacity for at least `additional` more
+    /// [`FamStruct::Entry`](trait.FamStruct.html#associatedtype.Entry) elements.
+    /// If the capacity is already reserved, this method doesn't do anything.
+    /// If not this will trigger a reallocation of the underlying buffer.
+    ///
+    fn reserve(&mut self, additional: usize) {
+        let desired_capacity = self.len() + additional;
+        if desired_capacity <= self.capacity() {
+            return;
+        }
+
+        let current_mem_allocator_len = self.mem_allocator.len();
+        let required_mem_allocator_len = FamStructWrapper::<T>::mem_allocator_len(desired_capacity);
+        let additional_mem_allocator_len = required_mem_allocator_len - current_mem_allocator_len;
+
+        self.mem_allocator.reserve(additional_mem_allocator_len);
+    }
+
+    /// Updates the length of `self` to the specified value.
+    /// Also updates the length of the `T` structure and of `self.mem_allocator` accordingly.
+    /// Reserves additional capacity if the len is increased.
+    /// Deallocates unnecessary memory if the len is decreased.
+    ///
+    /// This method might trigger reallocations of the underlying buffer.
+    ///
+    /// # Error: When len is greater than the max possible len it returns Error::SizeLimitExceeded
```

Comment:
This will look a bit weird in markdown. You basically have a heading title that is very very long.

Please use as heading title `Errors` as this is the standard defined in the Rust documentation guidelines (even if the function can return just on type of error). The explanation for the error needs to be in a paragraph.